### PR TITLE
[Storage Management] Metadata of Storage in storage.json

### PR DESCRIPTION
**Pull request by janlemcke**
# Introduction
This PR implements the caching support of TPCH using the mmap-based chunks. In order to do so, a storage.json file is created to store meta data about each table. To run TPCH using the new mmap-based approach use `./hyriseBenchmarkTPCH -s 0.01 --use_mmap` .

# Storage.json format:
```
{
    {
    "customer": {
        "chunk_count": 1,
        "column_count": 8,
        "columns": [
            {
                "column_name": "c_custkey",
                "data_type": "Int",
                "nullable": false
            },
           ...
        ],
        "file_count": 1
    },
    ...
}
```

# File structure:

Bildschirm­foto 2023-03-13 um 13 18 27

### CODE REVIEW

**Comment by Bouncner** [comment]
1.: Not sure if I like the two-way connection. Let's discuss on Monday. Good point.
2.: Good point, I'd say chunks can be finalized whenever somebody wants them to be finalized (might be the case in many tests). If the table or chunk is not known to the storage manager, nothing needs to happen.
3. When we talked about the JSON a couple of weeks ago, the  main points have been: (i) anything that's regularly read by users and which would be a pain to read from a binary file (that doesn't say a lot, but I am sure we'll later find such information) and (ii) information that should be stored again and again and again. For the tables, that's the table names, which files belong to a given table, and which columns the table stores (at least data types and names). Real world systems often use table names which cannot be simply used as the file names (SAP systems have tables with a `/` in the name). The columns do not need to be repeating for every chunk/file (even though it wouldn't really hurt).

**Comment by rrcomtech** [comment]
@janlemcke In the storage manager discussion the following was hinted:

> wenn wir für die Query-Ausführung speichermediumsabhängige Optimierungen durchführen wollen, wäre es hilfreich, pro Segment zu wissen, auf welche Art von Speicher dieses liegt.

Could you add this to the implementation? (For the moment, a dummy impl would be sufficient, in case it is not as easy to find it out)

**Comment by Bouncner** [comment]
That shouldn't be part of the json though. Wrong issue, @rrcomtech ?

**Comment by janlemcke** [comment]
### Stored Information

Information about the columns per table were added.

```
"columns": [
  {
      "data_type": 1,
      "id": 0,
      "name": "p_partkey"
  },
  ... ,
]
```

### How to connect chunk & table in StorageManager?

To iterate faster, i added two public fields to the chunks:
```
ChunkID _chunk_id;
std::string _table_name;
```
The `_chunk_id` is needed to uniquely identify the chunk within the storage.json for possible changes in the future during runtime. The `std::string _table_name` is needed to know to which table the chunk belongs to. 

Previously, the names for each table are only given in the map attribute of the storage manager. Now, the table itself must be aware of its name to propagate this information towards its chunks. This propagation is done during the append method or the table constructor. A table gets a "valid" its name during the add_table method of the SM.

### How could we build this differently?
To avoid a new class variable on chunk level, we could propagate this information through the finalize method. I have currently decided against it out of convenience, because of many code snippets that would need to be adjusted. The finalize() function is being called from operators, tests & benchmarks.

### How was the TPCH adapted?
The TableBuilder class has a new method "get_table()" that returns the table. During the initial creation of the TPCH data, the tables are registered directly to SM. After registration, the tables are populated with data. This results in `finalize()` calls during the first few seconds of the TPCH runtime --> easy to develop.

**Comment by janlemcke** [comment]
Here is a short and not 100% detailed diagram showing how the individual parts are connected.
grafik

**Comment by phoeinx** [comment]
>To my mind, the storage.json contains meta-data about the persistent storage. Probably, it is not supposed to be put in the actual file, the life cycle of the storage.json would still determine this information. Therefore, I'd say it has to be kept in mind when thinking about the managing the storage.json.

@rrcomtech: I am not sure if I understand you correctly - is this discussing the storing of which kind of storage each chunk is stored on in the storage.json? As the memory used to store a chunk can change at run time, DRAM-stored chunks would never need to be stored in the storage.json and further I am not sure when we would benefit from being able to access (and consistently update!) this data in a file when storing it in memory would lead to far better access times, I am not sure why this might be necessary.

**Comment by phoeinx** [comment]
@janlemcke  thank you for your work and nice that it's already that progressed!

Reading the PR and its discussion I did realize that I previously had a different understanding of what the `storage.json` was supposed to accomplish. To make this clear for me I would first like to ask/discuss a pretty basic question:

What exactly is the job of the `storage.json`?

From what I've read there is two possible interpretations (please add/correct what I didn't get right):

A. Enabling a cold start by providing a list of tables, their necessary meta-data and references to all "chunk"-files that store their data. 

B. Storing every chunk and its metadata as well its current storage location (as is the current implementation).

I did read @Bouncner s comment that the storage.json should also store easily accessible, human-readable data which would be accessed often and for which we want to provide more convenience than binary files. But what data would we need there exactly? 

I currently see a lot of duplication between our binary files storing the data and the storage.json. The binary files with their header structure are readable and processable as a single unit without requiring metadata. The storage.json stores a lot of information duplicatedly (or the binary files store it duplicatedly) - e.g. the row count of a chunk, the chunk_id etc..

I don't think this duplication makes sense for our current implementation: either we should store less information in the storage.json / only necessary information to obtain all tables, their structure and binary files storing their data and use the self-contained binary-files for the rest 
or
we shouuld remove some information from our binary files and use the storage.json as metadata to parse them. 

I am also wondering if storing that much data (which might update at various points (row_count, mutability, added chunks etc.) makes sense in the storage.json? We would have a single file that needs a lot of updates (and locks etc.). 

In conclusion I am currently a bit confused to the purpose of the storage.json (and especially to the discussion whether we should store metadata of each chunk in it), and would really appreciate hearing your opinions.

**Comment by Bouncner** [comment]
> @janlemcke thank you for your work and nice that it's already that progressed!
> 
> Reading the PR and its discussion I did realize that I previously had a different understanding of what the `storage.json` was supposed to accomplish. To make this clear for me I would first like to ask/discuss a pretty basic question:
> 
> What exactly is the job of the `storage.json`?
> 
> From what I've read there is two possible interpretations (please add/correct what I didn't get right):
> 
> A. Enabling a cold start by providing a list of tables, their necessary meta-data and references to all "chunk"-files that store their data.
> 
> B. Storing every chunk and its metadata as well its current storage location (as is the current implementation).

I cannot give you a full answer to be honest. We have learned over the past years, that handling binary data can be quite cumbersome and readable JSON files are much easier to handle in this case. So that brought up the idea of such a JSON file.
It is definitely used for purpose A.
Concerning B, I don't know what "storing every chunk" means. It might store some duplicated data (what would come to my mind is the set of ChunkIDs per file. But apart from that it should mainly avoid duplicated data (storing column types and names in every binary chunk would be an unnecessary overhead).

**Comment by rrcomtech** [comment]
@phoeinx I think the idea behind it was to be able to optimize on HW level. It was only mentioned once by @mweisgut and I guess we don't have to further discuss it atm.

File: src/lib/storage/storage_manager.hpp [hunk 1]
```diff
@@ -6,13 +6,13 @@
...
 #include <map>
 #include <memory>
-#include <nlohmann/json.hpp>
 #include <shared_mutex>
 #include <string>
 #include <vector>
 
 #include "chunk.hpp"
 #include "lqp_view.hpp"
+#include "nlohmann/json.hpp"
```
Comment (Bouncner) [inline src/lib/storage/storage_manager.hpp:15]: Separate third party from own headers.

**Comment by Bouncner** [comment]
Can you show us a short example of the current JSON?

**Comment by janlemcke** [comment]
Hey, thanks to all for the valuable feedback. In fact, because my understanding was to be able to do A (cold start), the code in fact serializes and deserializes the `_tables_current_persistence_file_mapping` map (from https://github.com/hyrise-mp-22-23/hyrise/pull/117) and incorporates additional data, such as column_count, columns etc.

This is a snippet from the current json:

```

{
    "table_files_mapping": [
        {
            "column_count": 5,
            "columns": [
                {
                    "data_type": 1,
                    "id": 0,
                    "name": "ps_partkey"
                },
                {
                    "data_type": 1,
                    "id": 1,
                    "name": "ps_suppkey"
                },
                {
                    "data_type": 1,
                    "id": 2,
                    "name": "ps_availqty"
                },
                {
                    "data_type": 3,
                    "id": 3,
                    "name": "ps_supplycost"
                },
                {
                    "data_type": 5,
                    "id": 4,
                    "name": "ps_comment"
                }
            ],
            "current_chunk_count": 0,
            "file_index": 0,
            "file_name": "partsupp_0.bin",
            "name": "partsupp",
            "row_count": 0
        },
        {
            "column_count": 9,
            "columns": [
                {
                    "data_type": 1,
                    "id": 0,
                    "name": "p_partkey"
                },
                {
                    "data_type": 5,
                    "id": 1,
                    "name": "p_name"
                },
                {
                    "data_type": 5,
                    "id": 2,
                    "name": "p_mfgr"
                },
                {
                    "data_type": 5,
                    "id": 3,
                    "name": "p_brand"
                },
                {
                    "data_type": 5,
                    "id": 4,
                    "name": "p_type"
                },
                {
                    "data_type": 1,
                    "id": 5,
                    "name": "p_size"
                },
                {
                    "data_type": 5,
                    "id": 6,
                    "name": "p_container"
                },
                {
                    "data_type": 3,
                    "id": 7,
                    "name": "p_retailsize"
                },
                {
                    "data_type": 5,
                    "id": 8,
                    "name": "p_comment"
                }
            ],
            "current_chunk_count": 0,
            "file_index": 0,
            "file_name": "part_0.bin",
            "name": "part",
            "row_count": 0
        },
}
```

File: src/benchmarklib/abstract_table_generator.cpp [hunk 2]
```diff
@@ -283,12 +283,16 @@ void AbstractTableGenerator::generate_and_store() {
 
       const auto add_table = [&]() {
         Timer per_table_timer;
+        auto output = std::string{""};
         if (storage_manager.has_table(table_name)) {
```
[the remark this answers is not in the document]
Comment (janlemcke) [reply src/benchmarklib/abstract_table_generator.cpp:287]: Good question. However, this code adaption within `tpch_table_generator.cpp` only resulted because i needed a (quick & dirty) way to somehow trigger the chunk::finalize method or at least gain knowledge about it. It is now obsolete and will be changed.

**Comment by Bouncner** [comment]
> 



> Hey, thanks to all for the valuable feedback. In fact, because my understanding was to be able to do A (cold start), the code in fact serializes and deserializes the `_tables_current_persistence_file_mapping` map (from #117) and incorporates additional data, such as column_count, columns etc.
> 
> This is a snippet from the current json:
> 
> ```
> 
> {
>     "table_files_mapping": [
>         {
>             "column_count": 5,
>             "columns": [
>                 {
>                     "data_type": 1,
>                     "id": 0,
>                     "name": "ps_partkey"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 1,
>                     "name": "ps_suppkey"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 2,
>                     "name": "ps_availqty"
>                 },
>                 {
>                     "data_type": 3,
>                     "id": 3,
>                     "name": "ps_supplycost"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 4,
>                     "name": "ps_comment"
>                 }
>             ],
>             "current_chunk_count": 0,
>             "file_index": 0,
>             "file_name": "partsupp_0.bin",
>             "name": "partsupp",
>             "row_count": 0
>         },
>         {
>             "column_count": 9,
>             "columns": [
>                 {
>                     "data_type": 1,
>                     "id": 0,
>                     "name": "p_partkey"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 1,
>                     "name": "p_name"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 2,
>                     "name": "p_mfgr"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 3,
>                     "name": "p_brand"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 4,
>                     "name": "p_type"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 5,
>                     "name": "p_size"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 6,
>                     "name": "p_container"
>                 },
>                 {
>                     "data_type": 3,
>                     "id": 7,
>                     "name": "p_retailsize"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 8,
>                     "name": "p_comment"
>                 }
>             ],
>             "current_chunk_count": 0,
>             "file_index": 0,
>             "file_name": "part_0.bin",
>             "name": "part",
>             "row_count": 0
>         },
> }
> ```

Looks good, a few comments:
 - what is `row_count` supposed to do?
 - can we find a better name for `name`? It's  the table name, right?
 - would we store the columns all over again and again for each file? (for SF 1.000, we will end up with many thousands of chunks!)
 - any reason to not use strings for `data_type`?

**Comment by mweisgut** [comment]
> Hey, thanks to all for the valuable feedback. In fact, because my understanding was to be able to do A (cold start), the code in fact serializes and deserializes the `_tables_current_persistence_file_mapping` map (from #117) and incorporates additional data, such as column_count, columns etc.
> 
> This is a snippet from the current json:
> 
> ```
> 
> {
>     "table_files_mapping": [
>         {
>             "column_count": 5,
>             "columns": [
>                 {
>                     "data_type": 1,
>                     "id": 0,
>                     "name": "ps_partkey"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 1,
>                     "name": "ps_suppkey"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 2,
>                     "name": "ps_availqty"
>                 },
>                 {
>                     "data_type": 3,
>                     "id": 3,
>                     "name": "ps_supplycost"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 4,
>                     "name": "ps_comment"
>                 }
>             ],
>             "current_chunk_count": 0,
>             "file_index": 0,
>             "file_name": "partsupp_0.bin",
>             "name": "partsupp",
>             "row_count": 0
>         },
>         {
>             "column_count": 9,
>             "columns": [
>                 {
>                     "data_type": 1,
>                     "id": 0,
>                     "name": "p_partkey"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 1,
>                     "name": "p_name"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 2,
>                     "name": "p_mfgr"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 3,
>                     "name": "p_brand"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 4,
>                     "name": "p_type"
>                 },
>                 {
>                     "data_type": 1,
>                     "id": 5,
>                     "name": "p_size"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 6,
>                     "name": "p_container"
>                 },
>                 {
>                     "data_type": 3,
>                     "id": 7,
>                     "name": "p_retailsize"
>                 },
>                 {
>                     "data_type": 5,
>                     "id": 8,
>                     "name": "p_comment"
>                 }
>             ],
>             "current_chunk_count": 0,
>             "file_index": 0,
>             "file_name": "part_0.bin",
>             "name": "part",
>             "row_count": 0
>         },
> }
> ```

Thanks for the comment. Some questions/comments:
- do we actually need the ID for each column? Can't we simply assume that the first array object's attribute belongs to the first column with ID 0, the second array object's attribute to the column with ID 1, and so on?
- what about nullability? Are all columns nullable by default?
- I agree with @Bouncner: having the data type as a string would be better human readable. You could get inspirations by viewing some `.csv.json` files in `/hyrise/resources/benchmark/tpcds/tables`.
- do we have (i) one json file per binary file or (ii) one json file per table? I assumed (ii). Therefore, we would have fields for specifying the table schema once and additionally information relevant for reading the chunks.
- Can "file_name" store a list of values, i.e., multiple file names? If not, does that mean that we have one json per binary file?
- what is file index used for?
- having an overview of what kind of metadata data is stored in a binary file would help decide what fields we want to store in the json file.

**Comment by mweisgut** [comment]
I think we can exclude row count since we also have it in the binary file.

File: src/benchmarklib/abstract_table_generator.hpp [hunk 3]
```diff
@@ -64,6 +64,8 @@ class AbstractTableGenerator {
 
   static std::shared_ptr<BenchmarkConfig> create_benchmark_config_with_chunk_size(ChunkOffset chunk_size);
 
+  bool cold_start_used = false;
```
Comment (phoeinx) [inline src/benchmarklib/abstract_table_generator.hpp:67]: would be really great to be able to pass this as an argument to the executable.
For our purposes now perhaps analogous to `_benchmark_config->cache_binary_tables` 
and later on something similarly for 
`./hyriseConsole --start-from-persisted-data <path_to_data>`

File: src/benchmarklib/benchmark_runner.cpp [hunk 4]
```diff
@@ -154,7 +154,7 @@ void BenchmarkRunner::run() {
     }
   }
 
-  _table_generator->delete_binaries();
+  //_table_generator->delete_binaries();
```
Comment (phoeinx) [inline src/benchmarklib/benchmark_runner.cpp:157]: This is probably commented out so that you can reuse the binaries right? Some nicer structure - with an extra directory for the persisted data files (binary) would be great.

**Comment by janlemcke** [comment]
FYI, current format:
```
{
    "customer": {
        "chunk_count": 1,
        "column_count": 8,
        "columns": [
            {
                "column_name": "c_custkey",
                "data_type": 1,
                "nullable": false
            },
            ...
        ],
        "file_count": 1,
        "file_name": "customer_0.bin"
    },
   ...
}
```

**Comment by janlemcke** [comment]
I removed file_name in the new version, because the filenames have the format "[tablename]_[fileindex].bin". Because of the key of the given json object and the file_count, we can create all filenames during runtime.

```
{
"customer": {
        "chunk_count": 1,
        "column_count": 8,
        "columns": [
            {
                "column_name": "c_custkey",
                "data_type": 1,
                "nullable": false
            },
            ...
        ],
        "file_count": 1
    },
...
}
```

**Comment by janlemcke** [comment]
Regarding point 2 of the previously defined TODOs. I added a new command line parameter for the TPCH benchmark:
`--use_storage_json`

### Behaviour (4 cases)
`./hyriseBenchmarkTPCH -s 0.1` runs the TPCH benchmark in the old fashion. If no binaries are already cached, it creates binaries and stores them into `tpch_cached_tables/sf-0.1` for sf 0.1 (**case 1**). If binaries for this scalefactor are available, it loads them instead of generating the tables from scratch (**case 2**)


`./hyriseBenchmarkTPCH -s 0.1 --use_storage_json` runs the TPCH benchmark with the newly implemented mmap-based chunks. 'Binaries' is now referring to binaries using the new file format. If no binaries are already cached (gets checked via the existence of a storage.json), it creates the table and in the end persists them. During the persistence the table gets automatically written to disk which in fact incorporates the caching (**case 3**). If a storage.json is found, instead of creating tables from scratch, the binaries are loaded using the storage.json and the persist() step is skipped, because the mapping between chunks and their data are already incorporateed (**case 4**).

File: src/lib/storage/storage_manager.cpp [hunk 5]
```diff
@@ -3,7 +3,9 @@
 #include <sys/fcntl.h>
 #include <sys/mman.h>
 #include <unistd.h>
+#include <fstream>
 #include <memory>
+#include <mutex>
```
Comment (tjjordan) [inline src/lib/storage/storage_manager.cpp:8]: Needed?

File: src/lib/storage/storage_manager.cpp [hunk 6]
```diff
@@ -24,6 +27,7 @@
 #include "storage/value_segment.hpp"
 #include "utils/assert.hpp"
 #include "utils/meta_table_manager.hpp"
+using json = nlohmann::json;
```
Comment (tjjordan) [inline src/lib/storage/storage_manager.cpp:30]: `nlohmann::json` is ok.

File: src/lib/storage/storage_manager.cpp [hunk 7]
```diff
@@ -638,8 +667,82 @@ PersistedSegmentEncodingType StorageManager::resolve_persisted_segment_encoding_
...
+      };
+      columns_json.push_back(column_object);
+    }
+    table_json["columns"] = columns_json;
+
+    _storage_json[mapping.first] = table_json;
+  }
+}
+
+void StorageManager::save_storage_json_to_disk() {
```
Comment (tjjordan) [inline src/lib/storage/storage_manager.cpp:695]: `update_storage_json`?

File: src/lib/storage/storage_manager.cpp [hunk 8]
```diff
@@ -638,8 +667,82 @@ PersistedSegmentEncodingType StorageManager::resolve_persisted_segment_encoding_
...
 }
 
-void StorageManager::export_compressed_vector(const CompressedVectorType type, const BaseCompressedVector& compressed_vector,
-                              std::ofstream& ofstream) {
+void StorageManager::_serialize_table_files_mapping() {
+  for (const auto& mapping : _tables_current_persistence_file_mapping) {
+    const auto table = get_table(mapping.first);
+    const auto column_count = table->column_count();
+    const auto chunk_count = mapping.second.file_index * MAX_CHUNK_COUNT_PER_FILE + mapping.second.current_chunk_count;
+    json table_json = {{"file_count", mapping.second.file_index + 1},
```
Comment (tjjordan) [inline src/lib/storage/storage_manager.cpp:675]: auto to stick

File: src/lib/storage/storage_manager.cpp [hunk 9]
```diff
@@ -638,8 +667,82 @@ PersistedSegmentEncodingType StorageManager::resolve_persisted_segment_encoding_
...
+
+void StorageManager::save_storage_json_to_disk() {
+  _serialize_table_files_mapping();
+  std::ofstream output_file(_cache_directory + _storage_json_name, std::ios::trunc);
+  const auto json_serialized = _storage_json.dump(4);
+  output_file << json_serialized;
+  output_file.close();
+}
+
+std::vector<TableColumnDefinition> StorageManager::get_table_column_definitions_from_json(
```
Comment (tjjordan) [inline src/lib/storage/storage_manager.cpp:703]: private?

File: src/benchmarklib/abstract_table_generator.cpp [hunk 10]
```diff
@@ -448,4 +446,48 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
   return table_info_by_name;
 }
 
+std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_load_binary_tables_from_json() {
+  std::unordered_map<std::string, BenchmarkTableInfo> table_info_by_name;
+  auto& storage_manager = Hyrise::get().storage_manager;
+  const auto tables_files_mapping = storage_manager.get_tables_files_mapping();
```
Comment (tjjordan) [inline src/benchmarklib/abstract_table_generator.cpp:452]: This should be included in the time calculation.

File: src/benchmarklib/abstract_table_generator.cpp [hunk 11]
```diff
@@ -448,4 +446,48 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
   return table_info_by_name;
 }
 
+std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_load_binary_tables_from_json() {
```
Comment (tjjordan) [inline src/benchmarklib/abstract_table_generator.cpp:449]: Put this into the AbstractTableGenerator.

**Review by tjjordan — COMMENTED** [review]
On the whole, the structure is appropriate.
Some improvements to the OOP need to be made (e.g., giving the Chunks the ability to persist themselves, reducing the file name passing).
Also, the old binary saving methods should be removed since they are now pretty redundant with our approach.

File: src/lib/storage/storage_manager.cpp (hunk 9 above)
Comment (janlemcke) [reply src/lib/storage/storage_manager.cpp:703]: Cannot be private because this function is called within the abstractTableGenerator to recreate the tables.


```
465 // Create column definitions
466   const auto column_definitions = storage_manager.get_table_column_definitions_from_json(table_name);
```

File: src/benchmarklib/abstract_table_generator.cpp (hunk 10 above)
Comment (janlemcke) [reply src/benchmarklib/abstract_table_generator.cpp:452]: What does this mean? In which time calculation?

File: src/benchmarklib/abstract_table_generator.cpp (hunk 11 above)
Comment (janlemcke) [reply src/benchmarklib/abstract_table_generator.cpp:449]: You mean into the StorageManager? It's already in the AbstractTableGenerator.

Comment (tjjordan) [reply src/benchmarklib/abstract_table_generator.cpp:449]: Oops, I'm sorry.
